### PR TITLE
Test to ensure admins can edit/update locked pages , closes #397

### LINKED ISSUE #397: Wiki page locking
## Overview

For extremely cautious use, for around 10 or so "most visible" pages on the site and/or those related to the staff only (board, for example), we need some locking abilities. Pages like:

- https://publiclab.org/wiki/board (has been hit a lot)
- https://publiclab.org/about
- https://publiclab.org/getting-started

Admins (`user.role == "admin"`) would have ability to lock, and moderators to edit locked pages. We'd show a clear message indicating that it's locked and why, and a way to appeal this, probably by linking to now-draft https://publiclab.org/wiki/moderation. in an alert. 

**Update:** as of #1095, tag creation permissions are done and we now just have to disallow the `edit` and `update` actions on https://github.com/publiclab/plots2/blob/master/app/controllers/wiki_controller.rb -- See **Tasks**, below.

Goal is to keep locking to an absolute minimum -- only for the most trafficked and spam-hit pages. Wiki page edit notifications for anything else. (#396)

****

## Tasks

1. [x] Maybe display a lock icon, with Bootstrap tooltip explanation, where Edit button is. (#1099)
2. [x] Also, link to a listing of all locked pages, at `/tag/locked`, from an explanation page at https://publiclab.org/wiki/locked (that'll be done in #1099)
3. [x] And an entry in the power tags page to go to `/wiki/locked`: https://publiclab.org/wiki/power-tags
4. [x] Intercept requests to the `edit` page. Look to see if the "locked" tag exists, and if it does, redirects back to the wiki page with an error message, by adding just after this line: https://github.com/publiclab/plots2/blob/master/app/controllers/wiki_controller.rb#L74 as follows  (complete in #1114):

```ruby
if @node.has_power_tag('locked') && (current_user.role != "admin" && current_user.role != "moderator")
  flash.now[:warning] = "This page is <a href='/wiki/power-tags#Locking'>locked</a>, and only <a href='/wiki/moderators'>moderators</a> can edit it."
  redirect_to @node.path
end
```

5) [x] (closed in #1155) We'd want to add the same kind of catch to `def update` in the same controller (for when you click `save` on a wiki page), except that we'd want to put all of these lines in an `else` block, so they don't get run if there's a lock:

https://github.com/publiclab/plots2/blob/master/app/controllers/wiki_controller.rb#L137-L173

* [x] (closed in #1155) Add the tag `locked` to a page in a test, and make sure a normal user can't edit that page. A test of a wiki page (example: https://github.com/publiclab/plots2/blob/master/test/functional/wiki_controller_test.rb#L124-L133) looks like this: 

```ruby
  test "basic user blocked from updating a locked wiki page" do
    node(:organizers).add_tag('locked', rusers(:admin)) # lock the page with a tag

    # then try updating it
    post :update,
         id:  node(:organizers).id, 
         uid:   rusers(:bob).id, 
         title: ""

    assert_template "wiki/show"
    assert_select ".alert", "expected message"
  end
```

* [x] follow the steps in the last task to add a test for the "edit" action (in #1114)
* [ ] test that admins can indeed access both of the pages above

This issue needs breaking up!

### PR DESCRIPTION
**Pull request by 500swapnil**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue
* [ ] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer
* [ ] `schema.rb.example` has been updated if any database migrations were added

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

**Comment by 500swapnil** [comment]
I made the changes in the link but cannot understand what test to add in wiki_controller_test.rb to ensure admin isn't redirected. Would like you to help with this one! @jywarren

File: app/views/wiki/show.html.erb [hunk 1]
```diff
@@ -43,7 +43,7 @@
       <ul class="nav nav-tabs">
         <li class="active"><a href="#"><span class="hidden-xs"><%= t('wiki.show.view') %></span><span class="visible-xs"><i class="fa fa-file-o"></i></span></a></li>
         <% if @node.has_tag('locked') %>
-          <li><a href="/wiki/locked"><i class="fa fa-lock"></i></a></li>
+          <li><a href="/wiki/edit/pagetitle"><i class="fa fa-lock"></i></a></li>
```
Comment (jywarren) [inline app/views/wiki/show.html.erb]: Oops, sorry -- `pagetitle` will actually have to be `<%= @node.edit_path %>?t=<%= Time.now.to_i %>` -- i was typing too fast and wasn't clear enough. Make sense?

Comment (jywarren) [inline app/views/wiki/show.html.erb]: And we should add a conditional -- `<% if current_user && (current_user.role == "moderator" || current_user.role == "admin") %>` with the "real" link, but if not, go to `/wiki/locked` -- make sense?

Comment (500swapnil) [inline app/views/wiki/show.html.erb]: I added a check. It serves the same purpose. Is it ok?

**Comment by 500swapnil** [comment]
@jywarren I added the check. It redirects if page is locked and user isn't an admin.

File: app/views/wiki/show.html.erb [hunk 2]
```diff
@@ -27,23 +27,23 @@
...
           | <a href="/n/<%= @node.id %>"><%= t('wiki.show.shortlink') %></a>
         </span>
       </div>
- 
+
       <ul class="nav nav-tabs">
         <li class="active"><a href="#"><span class="hidden-xs"><%= t('wiki.show.view') %></span><span class="visible-xs"><i class="fa fa-file-o"></i></span></a></li>
-        <% if @node.has_tag('locked') %>
-          <li><a href="/wiki/locked"><i class="fa fa-lock"></i></a></li>
+        <% if @node.has_tag('locked') && current_user && (current_user.role != "moderator" && current_user.role != "admin")%>
```
Comment (jywarren) [inline app/views/wiki/show.html.erb]: I'm sorry, actually this doesn't account for if there is no logged-in user... they'd see no lock symbol. So we need to add something like:

```ruby
        <% if @node.has_tag('locked') && (current_user ? (current_user.role != "moderator" && current_user.role != "admin") : true) %>
```

This uses the ternary operator, and I think it's correct, but if you can think of a simpler way to write it, I'm open to suggestions!

Comment (500swapnil) [inline app/views/wiki/show.html.erb]: Yeah this is a better method. Thanks!

**Comment by 500swapnil** [comment]
@jywarren Made the required changes. Please have a look!

**Comment by jywarren** [comment]
Oh, looks good but could you correct the indentation? Thanks, that should be it!

**Comment by 500swapnil** [comment]
Done!

**Comment by jywarren** [comment]
Great work!

**Comment by 500swapnil** [comment]
@jywarren Thanks! Any other issue you feel I can work with?

**Comment by jywarren** [comment]
Sure -- perhaps you could tackle some from the Geographic features? 

https://github.com/publiclab/plots2/issues/1070

Just be sure to coordinate with Mridul, who's also tackling those. 

This one is complex, but there is example code:  Placename autocompletion (break me up) like MapKnitter new map form, as done in this interface, which you can view at https://publiclab.org/profile/info/your_username


and it wouldn't overlap with Mridul's work, I think!

**Comment by 500swapnil** [comment]
@jywarren If he's already working with those, perhaps I should work with some other issue?

**Comment by jywarren** [comment]
Sure -- well, would you like to work on something more infrastructural and
database-oriented, or something more interface oriented?

**Comment by 500swapnil** [comment]
@jywarren I will co-ordinate with Mridul and work on Geographic features.

**Comment by jywarren** [comment]
OK, super!

**Comment by 500swapnil** [comment]
@jywarren But are there any interface oriented issues for now? Until I learn to work on geographic features. I can work on interface or database oriented issues. Thanks a lot!!

**Comment by jywarren** [comment]
Yes, there is this one:

https://github.com/publiclab/plots2/issues/904

It may involve integration of the rich editor, at
https://github.com/publiclab/PublicLab.Editor/

similarly to how it was done here:
https://github.com/publiclab/plots2/tree/master/app/views/editor/rich.html.erb

**Comment by 500swapnil** [comment]
@jywarren Thanks!

**Pull request merged**